[B + C] FoodLevelChangeEvent improvements - Adds BUKKIT-3581

gmcferrin opened the pull request:
a) This PR will allow developers to get the saturation level from FoodLevelChangeEvents, instead of only the food level. Currently, while it's possible to get / set the saturation level for a player, it's not possible to know how a given food changes the saturation.

b) Just like being able to work with the food level, being able to work with the saturation level is important for plugins that modify this event, as the raw food level / hunger is only one part of the food mechanic, and without having the saturation value it's impossible to completely alter the vanilla behavior.

c) This would not break existing plugins working from the Bukkit API, as the old FoodLevelChange event would still exist, but would return 0 for the saturation level. The raw CB event, however, would change, but it has long been said that relying on CB code to remain constant is unreliable, and breaks rather consistently.

CraftBukkit PR: https://github.com/Bukkit/CraftBukkit/pull/1078

turt2live commented:
Please follow the [Contributing Guidelines](https://github.com/Bukkit/Bukkit/blob/master/CONTRIBUTING.md#pull-request-formatting-expectations) for PR descriptions.

GuntherDW commented:
This PR was made before those guidelines were in place, so he'd have the know in advance that there would be such guidelines.

Nowadays he should have to update the description though indeed.

turt2live commented:
I understand that, that's why I linked them.

All PRs //should// have the same format regardless of age.

GuntherDW commented:
Most likely the author "forgot" about this PR though if he didn't update it, or didn't think he'd need to.

gmcferrin commented:
Guys...

Ever think I maybe just haven't gotten around to it?

Sincerely,
A really busy developer

turt2live commented:
I did not mean to put a timeline onto you. I understand how busy we all can be :)

gmcferrin commented:
@turt2live I know. Just letting you and @GuntherDW know that I hadn't forgotten about it.

GuntherDW commented:
Sorry if it felt like that @gmcferrin I didn't mean it like that.
I do know just how busy a developer can be, even though I'm more of a sysop by degree.

turt2live commented:
@gmcferrin If you are still interested in this please update it to the latest code base and resolve the mentioned concerns above. (friendly reminder, it has been 2 months since your last one...)

gmcferrin committed:
Improve FoodLevelChangeEvent. Adds BUKKIT-3581

Previously, FoodLevelChangeEvent only provided the player involved in the
event and the new food level of the player. This changes the event to also
provide the type of food being consumed, along with the new saturation and
exhaustion levels of the player.

gmcferrin commented:
@turt2live Updated. :D

turt2live commented:
@gmcferrin Friendly poke another 2 months later to update this :)

Minor things on the CraftBukkit side need addressing (mostly a rebase). Additionally, please take the time to update according to the contributing guidelines. 

Thanks!

turt2live commented on src/main/java/org/bukkit/event/entity/FoodLevelChangeEvent.java:
```diff
@@ -23,6 +35,15 @@ public HumanEntity getEntity() {
     }
 
     /**
+     * Get the food being eaten in this event.
```
`Gets`

turt2live commented on src/main/java/org/bukkit/event/entity/FoodLevelChangeEvent.java:
```diff
@@ -42,12 +63,70 @@ public int getFoodLevel() {
+
+    /**
+     * Sets the resultant exhaustion level that the entity involved in this event
+     * should be set to
```
There's a mention that Exhaution cannot exceed 40.0F, but that does not appear to be documented here.
